### PR TITLE
Rule S1751 Jump statements should not be used unconditionally

### LINKED ISSUE #214: Bug Detection Rule: "Jump statements should not be used unconditionally"
Implements [RSPEC-1751](https://jira.sonarsource.com/browse/RSPEC-1751)

### PR DESCRIPTION
**Pull request by valhristov**
Fix #214

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        internal const string DiagnosticId = "S1751";
+        private const string MessageFormat = "Remove this '{0}' statement or make it conditional.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<SyntaxKind> methodOrProperty = new HashSet<SyntaxKind>(new[]
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: You could use the inline initialization:
```
private static readonly ISet<SyntaxKind> methodOrProperty = new HashSet<SyntaxKind>
{
...
};
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        {
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.PropertyDeclaration,
+            SyntaxKind.ParenthesizedLambdaExpression,
+            SyntaxKind.SimpleLambdaExpression
+        });
+
+        private static readonly ISet<SyntaxKind> conditionaStatements = new HashSet<SyntaxKind>(new[]
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: Same as previous comment.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        {
+            SyntaxKind.IfStatement,
+            SyntaxKind.SwitchStatement,
+            SyntaxKind.CatchClause
+        });
+
+        private static readonly ISet<SyntaxKind> loopStatements = new HashSet<SyntaxKind>(new[]
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: Same as previous comment.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs (hunk 2 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: Fix the typo in the name of the variable.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs (hunk 1 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: I would rename this variable to specify syntax kind and the plural form (something like `methodOrPropertySyntaxKinds`).

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            context.RegisterSyntaxNodeActionInNonGenerated(c => 
+            {
+                var jumpStatement = (StatementSyntax)c.Node;
+
+                var statement = jumpStatement
+                    .Ancestors()
+                    .TakeWhile(IsWithinMethodPropertyOrLambda)
+                    .FirstOrDefault(IsConditionalOrLoop);
+
+                if (statement != null && IsLoop(statement))
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: You could have `IsLoop` handling null to make this call clearer.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/UnconditionalJumpStatement.cs [hunk 5]
```diff
@@ -0,0 +1,211 @@
+using System;
+
+namespace Tests.Diagnostics
+{
+    class Program
+    {
+        public void Test1(string[] strings)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/UnconditionalJumpStatement.cs]: Don't know what the rule should do if the `break` or `continue` is within an always true branch in a loop. Could you add a test?

Comment (valhristov) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/UnconditionalJumpStatement.cs]: I don't think we should test for that. Adding a test will mean that we duplicate the logic inside the always true/false rules, which will raise an issue anyway.
